tools/crm_mon.c

```diff
@@ -2107,26 +2107,23 @@ crm_diff_update(const char *event, xmlNode * msg)
 }
 
 static int
-get_fencing_history(stonith_history_t **stonith_history)
+get_fencing_history(stonith_t *st, stonith_history_t **stonith_history, bool reduce)
 {
-    int rc = 0;
+    int rc = pcmk_rc_ok;
 
-    while (fence_history) {
-        if (st != NULL) {
-            rc = st->cmds->history(st, st_opt_sync_call, NULL, stonith_history, 120);
+    if (st == NULL) {
+        rc = ENOTCONN;
+    } else {
+        rc = st->cmds->history(st, st_opt_sync_call, NULL, stonith_history, 120);
 
-            if (rc == 0) {
-                *stonith_history = stonith__sort_history(*stonith_history);
-                if (!pcmk_all_flags_set(show, pcmk_section_fencing_all)
-                    && (output_format != mon_output_xml)) {
+        rc = pcmk_legacy2rc(rc);
+        if (rc != pcmk_rc_ok) {
+            return rc;
+        }
```

Comment:
Primarily, I was interested in just rearranging it to make it a little easier to follow.  And then after that, I was thinking it would be nicer to have the caller decide whether it wants to retry or not, though I think that probably got lost in subsequent patches.  I'm not really sure that there is a great reason to keep retrying here aside from daemonize mode and maybe not even then.  I think for any use cases where we repeatedly show the status, we'll try to reconnect and re-fetch history every time.  And for one shot mode, do we really want to wait for fence history to be available?  I guess we might want to do that.

Reply:
I think the new approach is reasonable